### ISSUE TITLE
SU2 freezes for a case with exactly 4 cores

### ISSUE THREAD

**Issue by TobiKattmann** [su2code/SU2#1241]
Hi all,

When running `TestCases/incomp_navierstokes/streamwise_periodic/chtPinArray_2d/configMaster.cfg` with 4 cores, SU2 freezes after printing `Simulation Run using the Multizone Driver` to screen. No more screen or file output (e.g. history), but `top` shows that the processes are not idling and the respective cores are at ~100% workload. 

Call `mpirun -n 4 SU2_CFD configMaster.cfg` and tell me if it is the same for you. I tested on on our cluster and my private machine with the same result and it also happens every time. They seem to run forever  ... I let both setups run for an hour (on 3 cores it takes a couple of min to complete) without any change in what i see = no output + cores at 100%

With any other number !=4 for cores the code runs

When was it introduced? Not very very recent ... already there when #773 was merged 

**Now to the debugging efforts:**
Happens in debug and release mode.

I changed  quite a few options to see if sth interesting happens (no turb model, velocity inlet and pressure outlet, changed linear solvers and permutated values all over the place) but no change there. 

As it is just 4 cores I went attached to all processes with 
```
mpirun -n 4 xterm -e gdb --args SU2_CFD configMaster.cfg
```
And the backtraces show that 2 processes are waiting at:
<details>
  <summary>this place</summary>

```
#0  0x00002aaac3d5ec40 in hmca_bcol_basesmuma_bcast_k_nomial_knownroot_progress
    () from /opt/mellanox/hcoll/lib/hcoll/hmca_bcol_basesmuma.so
#1  0x00002aaabf5ee590 in hmca_coll_ml_allgather ()
   from /opt/mellanox/hcoll/lib/libhcoll.so.1
#2  0x00002aaabf32f7e0 in mca_coll_hcoll_allgather ()
   from /share/hpc_apps/openmpi/3.1.6/gnu/5.3.0/lib/openmpi/mca_coll_hcoll.so
#3  0x00002aaaaad2b34e in PMPI_Allgather ()
   from /share/hpc_apps/openmpi/3.1.6/gnu/5.3.0/lib/libmpi.so.40
#4  0x000000000062fc74 in CBaseMPIWrapper::Allgather (sendbuf=0x214f670, 
    sendcnt=1, sendtype=0x2aaaaafb4de0 <ompi_mpi_double>, recvbuf=0x213a400, 
    recvcnt=1, recvtype=0x2aaaaafb4de0 <ompi_mpi_double>, 
    comm=0x2aaaaafbbc60 <ompi_mpi_comm_world>)
    at ../SU2_CFD/src/output/filewriter/../../../include/output/filewriter/../../../../Common/include/parallelization/mpi_structure.hpp:220
#5  0x0000000000860117 in CSolver::SetResidual_RMS (this=0x21381c0, 
    geometry=0x21cb340, config=0x2073590)
    at ../SU2_CFD/src/solvers/CSolver.cpp:2183
#6  0x00000000007b6af8 in CHeatSolver::ImplicitEuler_Iteration (
    this=0x21381c0, geometry=0x21cb340, solver_container=0x23a1c10, 
    config=0x2073590) at ../SU2_CFD/src/solvers/CHeatSolver.cpp:1549
#7  0x0000000000b5a666 in CIntegration::Time_Integration (this=0x2381f70, 
    geometry=0x21cb340, solver_container=0x23a1c10, config=0x2073590, 
    iRKStep=0, RunTime_EqSystem=21)
---Type <return> to continue, or q <return> to quit---
    at ../SU2_CFD/src/integration/CIntegration.cpp:205
#8  0x00000000004a2973 in CSingleGridIntegration::SingleGrid_Iteration (
    this=0x2381f70, geometry=0x1fdc8b0, solver_container=0x1db9ad0, 
    numerics_container=0x1fddb10, config=0x1a6cb50, RunTime_EqSystem=21, 
    iZone=1, iInst=0)
    at ../SU2_CFD/src/integration/CSingleGridIntegration.cpp:73
#9  0x00000000004e33e4 in CHeatIteration::Iterate (this=0x218fdc0, 
    output=0x20c8630, integration=0x1a59a90, geometry=0x1fdc8b0, 
    solver=0x1db9ad0, numerics=0x1fddb10, config=0x1a6cb50, 
    surface_movement=0x1a6cc30, grid_movement=0x1a6cc50, FFDBox=0x1a6cc70, 
    val_iZone=1, val_iInst=0) at ../SU2_CFD/src/iteration/CHeatIteration.cpp:40
#10 0x00000000004db2ba in CFluidIteration::Solve (this=0x218fdc0, 
    output=0x20c8630, integration=0x1a59a90, geometry=0x1fdc8b0, 
    solver=0x1db9ad0, numerics=0x1fddb10, config=0x1a6cb50, 
    surface_movement=0x1a6cc30, grid_movement=0x1a6cc50, FFDBox=0x1a6cc70, 
    val_iZone=1, val_iInst=0)
    at ../SU2_CFD/src/iteration/CFluidIteration.cpp:302
#11 0x000000000047adec in CMultizoneDriver::Run_GaussSeidel (this=0x1fe0f80)
    at ../SU2_CFD/src/drivers/CMultizoneDriver.cpp:329
#12 0x000000000047a415 in CMultizoneDriver::StartSolver (this=0x1fe0f80)
    at ../SU2_CFD/src/drivers/CMultizoneDriver.cpp:193
#13 0x0000000000415413 in main (argc=2, argv=0x7fffffffbb18)
    at ../SU2_CFD/src/SU2_CFD.cpp:150
```
</details>
... and two other processes waiting at 
<details>
  <summary>that place</summary>

```
#0  0x00002aaac3d45500 in hmca_bcol_basesmuma_reduce_intra_fanin_progress ()
   from /opt/mellanox/hcoll/lib/hcoll/hmca_bcol_basesmuma.so
#1  0x00002aaabf587798 in _coll_ml_allreduce ()
   from /opt/mellanox/hcoll/lib/libhcoll.so.1
#2  0x00002aaabf33038d in mca_coll_hcoll_allreduce ()
   from /share/hpc_apps/openmpi/3.1.6/gnu/5.3.0/lib/openmpi/mca_coll_hcoll.so
#3  0x00002aaaaad2c20f in PMPI_Allreduce ()
   from /share/hpc_apps/openmpi/3.1.6/gnu/5.3.0/lib/libmpi.so.40
#4  0x000000000048a83f in CBaseMPIWrapper::Allreduce (sendbuf=0x7fffffff8a20, 
    recvbuf=0x1646158 <double CSysVector<double>::dot<CSysVector<double> >(VecExpr::CVecExpr<CSysVector<double>, double> const&) const::dotRes>, count=1, 
    datatype=0x2aaaaafb4de0 <ompi_mpi_double>, 
    op=0x2aaaaafc9920 <ompi_mpi_op_sum>, 
    comm=0x2aaaaafbbc60 <ompi_mpi_comm_world>)
    at ../SU2_CFD/src/drivers/../../include/drivers/../../../Common/include/parallelization/mpi_structure.hpp:205
#5  0x00000000004b0ef3 in CSysVector<double>::dot<CSysVector<double> > (
    this=0x23a20d0, expr=...)
    at ../SU2_CFD/src/integration/../../include/integration/../solvers/../../../Common/include/linear_algebra/CSysVector.hpp:317
#6  0x00000000004b0dcf in CSysVector<double>::squaredNorm (this=0x23a20d0)
    at ../SU2_CFD/src/integration/../../include/integration/../solvers/../../../Common/include/linear_algebra/CSysVector.hpp:331
---Type <return> to continue, or q <return> to quit---
#7  0x00000000004b0cf0 in CSysVector<double>::norm (this=0x23a20d0)
    at ../SU2_CFD/src/integration/../../include/integration/../solvers/../../../Common/include/linear_algebra/CSysVector.hpp:337
#8  0x0000000000da260b in CSysSolve<double>::BCGSTAB_LinSolver (
    this=0x23a2240, b=..., x=..., mat_vec=..., precond=..., 
    tol=1.0000000000000001e-15, m=20, residual=@0x7fffffff8fb8: 0, 
    monitoring=false, config=0x209bc10)
    at ../Common/src/linear_algebra/CSysSolve.cpp:553
#9  0x0000000000da404a in CSysSolve<double>::Solve (this=0x23a2240, 
    Jacobian=..., LinSysRes=..., LinSysSol=..., geometry=0x2105400, 
    config=0x209bc10) at ../Common/src/linear_algebra/CSysSolve.cpp:883
#10 0x00000000007b69da in CHeatSolver::ImplicitEuler_Iteration (
    this=0x23a1de0, geometry=0x2105400, solver_container=0x23a1d70, 
    config=0x209bc10) at ../SU2_CFD/src/solvers/CHeatSolver.cpp:1528
#11 0x0000000000b5a666 in CIntegration::Time_Integration (this=0x270d150, 
    geometry=0x2105400, solver_container=0x23a1d70, config=0x209bc10, 
    iRKStep=0, RunTime_EqSystem=21)
    at ../SU2_CFD/src/integration/CIntegration.cpp:205
#12 0x00000000004a2973 in CSingleGridIntegration::SingleGrid_Iteration (
    this=0x270d150, geometry=0x2005130, solver_container=0x20058c0, 
    numerics_container=0x20063b0, config=0x1de0e00, RunTime_EqSystem=21, 
    iZone=1, iInst=0)
    at ../SU2_CFD/src/integration/CSingleGridIntegration.cpp:73
---Type <return> to continue, or q <return> to quit---
#13 0x00000000004e33e4 in CHeatIteration::Iterate (this=0x270d170, 
    output=0x20f0cc0, integration=0x1e3efe0, geometry=0x2005130, 
    solver=0x20058c0, numerics=0x20063b0, config=0x1de0e00, 
    surface_movement=0x1de0e20, grid_movement=0x1de0e40, FFDBox=0x1de0e60, 
    val_iZone=1, val_iInst=0) at ../SU2_CFD/src/iteration/CHeatIteration.cpp:40
#14 0x00000000004db2ba in CFluidIteration::Solve (this=0x270d170, 
    output=0x20f0cc0, integration=0x1e3efe0, geometry=0x2005130, 
    solver=0x20058c0, numerics=0x20063b0, config=0x1de0e00, 
    surface_movement=0x1de0e20, grid_movement=0x1de0e40, FFDBox=0x1de0e60, 
    val_iZone=1, val_iInst=0)
    at ../SU2_CFD/src/iteration/CFluidIteration.cpp:302
#15 0x000000000047adec in CMultizoneDriver::Run_GaussSeidel (this=0x2009490)
    at ../SU2_CFD/src/drivers/CMultizoneDriver.cpp:329
#16 0x000000000047a415 in CMultizoneDriver::StartSolver (this=0x2009490)
    at ../SU2_CFD/src/drivers/CMultizoneDriver.cpp:193
#17 0x0000000000415413 in main (argc=2, argv=0x7fffffffbb18)
    at ../SU2_CFD/src/SU2_CFD.cpp:150
```
</details>
Look at the backtraces if you like. The first is further in the program (compare 1st number 6 and 2nd number number 10).
The first waits at an `Allgather` and the 2nd at an `Allreduce`. 

I am somewhat certain that the problem is this code bit which the second (the furhter back) backtrace points to. CSysVector.hpp -> `ScalarType dot` function l. 317
```c++
    /*--- Update shared variable with "our" partial sum. ---*/
    atomicAdd(sum, dotRes);
    SU2_MPI::Barrier(SU2_MPI::GetComm()); //added by tobi
#ifdef HAVE_MPI
    /*--- Reduce across all mpi ranks, only master thread communicates.
     * The nElm condition is to allow vectors to also be used locally. ---*/
    std::cout << "Rank: " << SU2_MPI::GetRank() << ", nElm != nElmDomain=" << (nElm != nElmDomain) << std::endl; //added by tobi
    if (nElm != nElmDomain) {
      SU2_OMP_BARRIER
      SU2_OMP_MASTER {
        sum = dotRes;
        const auto mpi_type = (sizeof(ScalarType) < sizeof(double)) ? MPI_FLOAT : MPI_DOUBLE;
        SelectMPIWrapper<ScalarType>::W::Allreduce(&sum, &dotRes, 1, mpi_type, MPI_SUM, SU2_MPI::GetComm());
      }
    }
#endif
    /*--- Make view of result consistent across threads. ---*/
    SU2_OMP_BARRIER

    return dotRes;
  }
```
The Problem is the `Allreduce` in the `if (nElm != nElmDomain) {` which leads to a freeze if not all ranks get the same result for that boolean.
The eagle eye prob already spotted the cout-Debugging in there. I also added a mpi barrier and the screen output i get is 
```
... only nElm != nElmDomain=1 above, which prob belong to the fluid iteration
Rank: 3, nElm != nElmDomain=1
Rank: 2, nElm != nElmDomain=1
Rank: 0, nElm != nElmDomain=1
Rank: 1, nElm != nElmDomain=1
Rank: 3, nElm != nElmDomain=1
Rank: 1, nElm != nElmDomain=1
Rank: 2, nElm != nElmDomain=1
Rank: 0, nElm != nElmDomain=0
Rank: 3, nElm != nElmDomain=0
```
and there it consequently freezes as two procs want to `Allreduce` and the other two never see that line of code. nElm and nElmDomain are not zero btw and have somewhat reasonable values in that case.
If I move 
```c++
const auto mpi_type = (sizeof(ScalarType) < sizeof(double)) ? MPI_FLOAT : MPI_DOUBLE;
SelectMPIWrapper<ScalarType>::W::Allreduce(&sum, &dotRes, 1, mpi_type, MPI_SUM, SU2_MPI::GetComm());
```
just outside the conditional then the code runs past that point and iterates normally. For that case it does not create any unwanted sideeffects. The explanatory comment
```
* The nElm condition is to allow vectors to also be used locally. ---*/
```
did not really enlighten me. I guess we have `nElm==nElmDomain` if we run without domain decomposition i.e. size=1. And as soon as size>1 `nElm` should be larger than `nElmDomain` as there is communication overhead ... at least how SU2 does it? So maybe the computation of those nElm values is not ok (I will look in that direction). On the other hand I guess such coding style (of putting mpi commands behind conditionals with local vars) provokes these errors? 

🆘 thanks already for the help and feel free to mention things I should try 

[This page](https://github.com/dealii/dealii/wiki/Frequently-Asked-Questions#how-do-i-debug-mpi-programs) with the 2 following sections was quite helpful.

**Desktop (please complete the following information):**
- gcc 5.3.0, openmpi 3.1.6, RHELS 7.6
- gcc 10.2.0, openmpi 4.0.5, Manjaro 21.0
 - SU2 Version: latest develop

**Comment by pcarruscag** [comment]
The intention was to allow CSysVector to be used like in a local way without mpi, because now it provides nice math syntax, (a+b-2*(c-a)) in an efficient way... But of course hell is full of good intentions and I guess if you have a disjoint domain, for example two things that do not touch, it is possible for that condition to evaluate differently on different ranks.
I'll remove it, thanks for the PI work.

**Comment by TobiKattmann** [comment]
Resolved by #1228  thanks @pcarruscag and thanks for the explanation. 

Just one more thing. do you have an idea why I still see 100% cpu load in that case? I picture a hyperactive process running around a stoplight sign at a zebra walk  🚥  🏃 🦓  waiting for his teammates to catch up whilst not getting cold. My private machine gets hot and loud so I guess there is real computation going on there. It could relax and sit on that park bench 🪑 for a minute.

**Comment by pcarruscag** [comment]
That is what imagine as well, i.e. very anxious waiting, some cpu might see that as a "power virus" and eventually slow down.
